File: rust/nucleotide_count.rs
use std::collections::HashMap;

fn is_nucleotide(nucleotide: char) -> bool {
    match nucleotide {
        'A' | 'C' | 'G' | 'T' => true,
        _ => false,
    }
}
pub fn count(nucleotide: char, dna: &str) -> Result<usize, char> {
    if !is_nucleotide(nucleotide) {
        return Err(nucleotide);
    }

    let mut ct = 0;

    for d in dna.chars() {
        if !is_nucleotide(d) {
            return Err(d);
        }
        if d == nucleotide {
            ct += 1;
        }
    }

    Ok(ct)
}

pub fn nucleotide_counts(dna: &str) -> Result<HashMap<char, usize>, char> {
    let mut mp = HashMap::new();

    "ACGT".chars().for_each(|c| {
        mp.insert(c, 0);
    });

    for d in dna.chars() {
        if !is_nucleotide(d) {
            return Err(d);
        }
        mp.entry(d).and_modify(|ct| *ct += 1);
    }

    Ok(mp)
}

File: lua/nucleotide-count.lua
local DNA = {}
DNA.__index = DNA

function DNA:new(s)
    if s:match("[^ATCG]") then
        return error("Invalid Sequence")
    end
    local t = { nucleotideCounts = { A = 0, T = 0, C = 0, G = 0 } }
    for c in s:gmatch(".") do
        t.nucleotideCounts[c] = t.nucleotideCounts[c] + 1
    end
    setmetatable(t, self)
    return t
end

function DNA:count(c)
    local ct = self.nucleotideCounts[c]
    assert(ct, "Invalid Nucleotide")
    return ct
end

return DNA
